libc/minimal: Add rand and srand

yashi opened the pull request:
rand() and srand() are pseudo-random number generator functions
defined in ISO C.  This implementation uses the Linear congruential
generator algorithm with the following parameters, which are the same
as GNU Libc.

  Modulus 0x7fffffff
  Multiplier 1103515245
  Increment 12345

ISO C / POSIX.1-2001 uses bits high 15 bits for Modulus, thus RAND_MAX
32767.  ref rand(3).

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi commented:
Or should I call `sys_rand32_get()`, instead?

yashi commented:
@enjiamai, @KangJianX would mind to give me some feedback?

enjiamai commented:
Hi @yashi ,  OK, please give me some time to take look at this, I'll feedback to you later. thanks!

enjiamai commented:
Hi @yashi , I have no problem with this implementation, it looks good, and I think we need this. I would so appreciate it if some basic test cases for this can also be added to tests/lib/c_lib/src/. How do you think? thank you!

enjiamai reviewed (changes requested):
Could you please help to add the test cases for these, thank you!

enjiamai reviewed (approved):
Looks great~ no problem from my side, thanks!

pfalcon commented on lib/libc/minimal/source/stdlib/rand.c:
```diff
@@ -0,0 +1,33 @@
+	}
+	__seed = s;
+}
+__weak FUNC_ALIAS(_srand, srand, void);
```
What's the reason for defining underscore version of a function, and then add a weak alias with normal name? (Why not use the normal name directly?)

yashi replied:
I followed other functions in minimal libc.  Why do they have aliases?

pfalcon replied:
> I followed other functions in minimal libc. Why do they have aliases?

Where do you see that?

```
zephyr/lib/libc/minimal/source$ grep -r -i weak *
zephyr/lib/libc/minimal/source$ grep -r -i alias *
```

That's exactly my point - existing sources in zephyr/lib/libc/minimal/source/ **don't** have any aliasing like that, and I'd suggest to not add it here either.

yashi replied:
Oops! Thanks.  I was checking `lib/libc/newlib/libc-hooks.c`.  Fixed.

pfalcon replied:
> Oops! Thanks. I was checking lib/libc/newlib/libc-hooks.c. Fixed.

I see. That's actually for newlib libc. Thanks for update!

pfalcon commented on lib/libc/minimal/source/stdlib/rand.c:
```diff
@@ -0,0 +1,33 @@
+#define MULTIPLIER (1103515245U)
+#define INCREMENT (12345U)
+
+static uint32_t __seed;
```
Another question: given that the function is defined in terms of standard C types (int/unsigned), then why uint32_t is used here?

pfalcon replied:
Final question: how rand() will behave, if srand() is not explicitly called, given such seed variable definition, and comment below `/* seed of 0 doesn't generate a good random number */` ?

yashi replied:
for `uint32_t`, it was just my habit to use size specific type when ever possible.  I'll change it to `unsigned int` if required.

Because the seed is static, it's `0` if `srand()` is not called.  Thus, it returns `1103527590`.

yashi replied:
oops, `rand()` without calling `srand()` means that `rand()` will use `0` as the seed and returns `12345` instead of `1103527590` (which is the value for the seed `1`).

hmm... I'm not sure we should guard seed of 0 or not.  Does anyone have any idea?

pfalcon replied:
> Does anyone have any idea?

Yes, it should be something like:

```
 static unsigned long srand_seed = 1; 
```

(Fixing also other issues like double-underscore prefix; variable type is based on sample implementation in https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html)

yashi replied:
fixed.

pfalcon commented on tests/lib/c_lib/src/main.c:
```diff
@@ -791,6 +791,49 @@ void test_time(void)
+ * @brief test srand functions
+ *
+ */
+void test_srand(void)
```
These tests look good, but ideally, the test would verify that rand() results after calling srand() with the same value are reproducible (per the definition of srand() semantics). I.e.: call srand() with one value, and check that a few (2-3) rand() calls give some values, then call srand() with another value, and call rand() a few times and check that values are different from 1st time. Then call srand() again with the initial value, and check that rand() give the same values as first time.

(That's in addition to the existing testcases - again, they're fine.)

yashi replied:
`zephyr/tests/lib/c_lib/README.txt` states:
> It is intended to catch issues in which a library is completely absent or non-functional, and is NOT intended to be a comprehensive test suite of all functionality provided by the libraries.

Regardless, I've added a reproducibility test for `rand()`.

pfalcon replied:
Sure, that was just a suggestion, as long as you went to write a more or less detailed test at all. I appreciate adding that test.

pfalcon reviewed (changes requested):
I'm always +1 for sustainably growing the minimal library. But IMHO there're few cleanups to be done and questions to be answered before this is ready for inclusion.

Another note - we usually try to put tests into a separate commit vs the core changes.

pfalcon commented:
I also find the current commit message to be confusing:

```
            initstate(0, state, sizeof(state));
            printf("%d\n", rand());

            return 0;
    }

See initstate(3p) for more detail.
```

You don't implement initstate(), so why people should see its man? I however followed the advice and what I see is:

```
       long int random(void);

       void srandom(unsigned int seed);

       char *initstate(unsigned int seed, char *state, size_t n);

       char *setstate(char *state);
```

In other words, rand/srand vs random/initstate seems to be two parallel, independent pseudo-random number generation function sets. You here don't implement the latter, so references to initstate() should rather be removed. Likewise, please double-check that rand/srand implementation matches that of glibc. If it is, please update the sample in the commit message to use rand/srand. And if it's not, I'd suggest to just use the sample implementation from POSIX: https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html and remove references to glibc.

(And if there's some confusion on my side, please let me know what I'm missing.)

yashi committed:
tests: c_lib: Add tests for rand() and srand()

 - simple tests for each function
 - a reproducibility test for rand with the same seed values

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi commented:
In order to test the Linear congruential generator algorithm, which is refered as `TYPE_0` in GLIBC, you have to call `initstate()` with [a state array of size 8 bytes][1] or [smaller than 32 byte][2].  Please note that the man page you referred is `3` not [`3p`, or `3posix`, from Linux man-pages project][2].

I've updated my commit log as below for more clarity.
> See initstate(3p) for more detail about how to use LCG in GLIBC.

[1]: https://www.redhat.com/en/blog/understanding-random-number-generators-and-their-limitations-linux
[2]: https://man7.org/linux/man-pages/man3/initstate.3p.html

pfalcon commented:
@yashi 

> In order to test the Linear congruential generator algorithm, which is refered as TYPE_0 in GLIBC, you have to call initstate() with a state array of size 8 bytes or smaller than 32 byte.

That's an interesting implementation detail of GLIBC, but I don't see how that relates to the functions srand()/rand() which are being implemented here. Nothing in https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html mentions initstate(). Nor you implement initstate(). So, why do you mention it at all?

Let's start from the beginning. Do you want Zephyr's implementation of srand()/rand() to have the same behavior/produce the same values as Glibc's srand()/rand()? If so, I'm generally +1 for that, as Glibc is a well known library and I'm for being consistent with well-know de-facto standard. But then please sure that you exactly follow behavior of srand()/rand() and don't involve unrelated implementation details, like the fact that in Glibc, srand() is implemented in terms of initstats().

And well, if you don't have that specific requirement to follow Glibc behavior, just use the simple algorithm from "Generating the Same Sequence on Different Machines" subheading of https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html .

> Please note that the man page you referred is 3 not 3p, or 3posix, from Linux man-pages project.

That page still described the functions:
```
       char *initstate(unsigned seed, char *state, size_t size);
       long random(void);
       char *setstate(char *state);
       void srandom(unsigned seed);
```

And not srand()/rand().

yashi commented:
@pfalcon, thanks for your feedback.  The reason why I've chosen GLIBC LCG is that I assumed most of the developers in the Zephyr community prefer to have the same algorithm as their development environment, which is, I also assume, Linux with GLIBC.  I don't have any preferences at all.  And the reason I put the reference in the commit log is that when someone wants to verify my implementation, it'd be nice to give a point or two for them.  I thought it'd be helpful for future references.  Nothing else.

I followed GLIBC LCG's _behavior_ in the Zephyr implementation.  No detail, nor `TYPE_0` thing is in Zephyr implementation.  The reference is just for a hint who wants verification in behavior.  If the default algorithm in GLIBC is LCG, I didn't put the reference in the commit log, but it's not.  If you just run `rand()` and `srand()` on your Linux, you are using a different algorithm and you get different values.  I'm not sure why the reference/doc in the commit log is such a big issue.

Again, I don't have any preferences in the algorithms.  I just want `rand()` and `srand()` for compatibility.  I can either remove the reference from the commit log or implement the C standard.  Please let me know which way you want.

pfalcon commented on lib/libc/minimal/source/stdlib/rand.c:
```diff
@@ -0,0 +1,27 @@
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <toolchain.h>
```
Is this include really needed?

yashi replied:
nop.  I'll leave `stdlib.h` there for checking the `rand()` and `srand()` declarations against the definitions.

pfalcon commented:
@yashi 

> The reason why I've chosen GLIBC LCG is that I assumed most of the developers in the Zephyr community prefer to have the same algorithm as their development environment, which is, I also assume, Linux with GLIBC.

I see, understand and appreciate such an approach. My concern is to make sure that if this approach is taken, then it actually works consistently, and described in a way to avoid confusion.

I now played with the sources myself, and took a closer look at glibc source (ref: https://code.woboq.org/userspace/glibc/stdlib/random.c.html#unsafe_state), and understood the situation: the default algorithm in glibc is "TYPE_3", while you implement "TYPE_0" here. Thus, to get compatible output from glibc, we need to switch it to TYPE using that initstate() call, after which both srand() and rand() will work in a way compatible with the implementation here.

I thus would recommend to update the commit message to explain that, e.g.:

> This implementation uses the Linear Congruential
Generator (LCG) algorithm with the following parameters, which are the
same as **used in** GNU Libc **"TYPE_0" algorithm**. **Note that the default
algorithm used by GNU Libc is not TYPE_0, and TYPE_0 should be selected
first by an initstate() call as shown below.**

> ...

    #include <stdio.h>
    #include <stdlib.h>

    int main()
    {
            static char state[8];

            /* Switch GLIBC to use LCG/TYPE_0 generator type. */
            initstate(0, state, sizeof(state));

            srand(1);  /* Or any other value. */
            printf("%d\n", rand());
            printf("%d\n", rand());

            return 0;
    }

pfalcon reviewed (approved):
I'd recommend to clarify situation re: GLIBC compatibility, as discussed in the previous comment.

Otherwise, looks good.

yashi committed:
doc: guides: Add srand()

Under the "OS Abstraction / POSIX Support" section in the "User and
Developer Guides", we somehow had rand() function already listed and
marked supported, but not srand().  Add srand() to the list.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi committed:
tests: crypto: mbedtls: Remove local implementation of rand()

When MBEDTLS_RSA_C is defined, mbedtls define its local version of
rand() function.  Since we already have rand() in our minimal libc, we
can safely remove this.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi commented:
It seems like I need to setup memory domain for static storage, `srand_seed`, to pass the tests from `tests/crypto/mbedtls`.  For newlib, [as documented here][1], there seems to be `z_libc_partition` but I'm not sure I can use it for minimal libc.

Would you please give me some hint how I should go with it?

[1]: https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#memory-domains

yashi committed:
doc: Markup kconfig options with :kconfig: role

Some Kconfig options are left marked as inline literals.  But in
Zephyr document, we use the "kconfig" role provided by Sphinx.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi committed:
doc: reference: memory_domain: Markup variables in sentences

Mark variables up if used in plain sentences for easier reading.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi committed:
libc: minimal: Add rand() and srand()

rand() and srand() are pseudo-random number generator functions
defined in ISO C. This implementation uses the Linear Congruential
Generator (LCG) algorithm with the following parameters, which are the
same as used in GNU Libc "TYPE_0" algorithm.

  Modulus 2^31
  Multiplier 1103515245
  Increment 12345
  Output Bits 30..0

Note that the default algorithm used by GNU Libc is not TYPE_0, and
TYPE_0 should be selected first by an initstate() call as shown below.

All global variables in a C library must be routed to a memory
partition in order to be used by user-mode applications when
CONFIG_USERSPACE is enabled.  Thus, srand_seed is marked as
such. z_libc_partition is originally used by the Newlib C library but
it's generic enough to be used by either the minimal libc or the
newlib.

All other functions in the Minimal C library, however, don't require
global variables/states.  Unconditionally using z_libc_partition with
the minimal libc might be a problem for applications utilizing many
custom memory partitions on platforms with a limited number of MPU
regions (eg. Cortex M0/M3). This commit introduces a kconfig option
CONFIG_MINIMAL_LIBC_RAND so that applications can enable the
functions if needed.  The option is disabled by default.

Because this commit _does_ implement rand() and srand(), our coding
guideline check on GitHub Action finds it as a violation.

    Error: lib/libc/minimal/include/stdlib.h:45:WARNING: Violation to
    rule 21.2 (Should not used a reserved identifier) - srand

But this is false positive.

The following is a simple test program for LCG with GNU Libc.

  #include <stdio.h>
  #include <stdlib.h>

  int main()
  {
          static char state[8];

          /* Switch GLIBC to use LCG/TYPE_0 generator type. */
          initstate(0, state, sizeof(state));

          srand(1);  /* Or any other value. */
          printf("%d\n", rand());
          printf("%d\n", rand());

          return 0;
  }

See initstate(3p) for more detail about how to use LCG in GLIBC.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi committed:
doc: reference: memory_domain: Update minimal libc

The global variables in the Minimal C library are now placed in
z_libc_partition, update the document accordingly.

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

yashi commented:
spent a few hours reading memory partitions and memory domains, and came up with the updated patch set. The following twister command gives me the same result as the current master 0c4e9d29bbc1b.  (all passed except for failed native_posix and skipped qemu_x86_tiny w/ both SDK 0.12.4 and 0.13.0 alpha 1).
```
$ scripts/twister -N -T tests/crypto/mbedtls -v
```
Hope I got it right.

pfalcon commented:
@yashi, Thanks for looking further into this. The situation reminds me my experience with hacking POSIX and/or libc subsystems, when seemingly simple changes leed to cascade of issues and follow-up changes. So, you're not alone ;-).

I looked into the situation myself. mps2_an385 supports 8 MPU regions, and one after another protection features enabled, that trickles into 3 "free" partition slots available for application code. Having additional libc partition indeed gets this down to 2 free partition slots, which causes the mem_protect/userspace test to fail.

I was able to get it pass with `CONFIG_TEST_HW_STACK_PROTECTION=n`. So, we could potentially make another test config for mps2_an385 with that option and exclude it from normal tests. I don't think that would scale, because 8 MPU regions is standard for Cortex-M0/Cortex-M3 devices, so the same problem would happen to many other (real) devices, the only special thing about mps2_an385 is that it's emulated via qemu, and we run this emulation during CI. Try to filter all affected devices would be too adhoc and cumbersome, and I guess we'd really like to run the rest with stack protection anyway.

So, I'd suggest to do partition plucking on where we add it now - on minimal libc size. It shouldn't be too bread-crumb configuration either, so instead of e.g. CONFIG_MINIMAL_LIBC_RAND, I'd suggest to add something like CONFIG_MINIMAL_LIBC_GLOBALS with a description like "Support C library functions which implicitly access global/static data, e.g. rand(). Such data must be put into a memory partition, and disabling this option may save an entry for application defining many custom partitions. ..."

Then disable this option for the affected test(s).

yashi commented:
@pfalcon, Thank you for your suggestion.  I've implemented what you suggested (in a way I understand, at least).   Please take a look the commit "libc: minimal: Use a memory partition iff globals are required".
Any feedback is as always welcome. :smiley:

pfalcon commented:
> Please take a look the commit "libc: minimal: Use a memory partition iff globals are required".
Any feedback is as always welcome.

Yes, that looks pretty much how I pictured it in my mind, but I'd suggest to not change  tests/kernel/mem_protect/userspace/testcase.yaml to add another test variant, bit instead just add "CONFIG_MINIMAL_LIBC_GLOBALS=n" to its prj.conf. After all, the whole idea of introducing CONFIG_MINIMAL_LIBC_GLOBALS was to avoid adding yet another test variant for mem_protect/userspace (which already has a few of them).

So, my suggestion would be to add to tests/kernel/mem_protect/userspace/prj.conf following:

```
# Disable libc globals partition to make the test runnable on devices
# with 8 MPU regions.
CONFIG_MINIMAL_LIBC_GLOBALS=n
```

pfalcon commented:
And another suggestion is to squash related commits together, e.g.:
* libc: minimal: Add rand() and srand()
* libc: minimal: Make srand_seed memory partition aware
* libc: minimal: Use a memory partition iff globals are required 

should be one; 2 of "tests: c_lib:" commits can also be squashed together.

pfalcon commented:
> Just noticed that in #26545, it's using LIBC_PARTITION instead of LIBC_PART as in this PR.

I guess LIBC_PART is fine. And that PR shows that it may be not easy to get bigger changes thru, as they may "pull strings" for even more changes, and some PRs get stale. So again, thanks, thanks for looking into related issues here, hopefully, it will make it easier to add other libc's/functions later.

yashi commented:
@pfalcon, I've updated the patchset.  Fixed the prj.conf instead of the testcase, merged "libc: minimal:" commits in one (merged all 4 because "libc: minimal: Place globals in z_libc_partition" causes a regression if you test it alone), merged two tests in one.

I'd like to merge "tests: crypto: mbedtls: Remove local implementation of rand()" into "libc: minimal: Add rand() and srand()" because without the fix, test for mbedtls failed to build.  WDYT?

pfalcon commented:
@yashi 

> I'd also like to sqiash "tests: crypto: mbedtls: Remove local implementation of rand()" into "libc: minimal: Add rand() and srand()" because without the fix, tests for mbedtls failed to build. WDYT?

Well, I'd personally err on the side of having clearly delinated commits rather than on having perfect bisectability (IMHO it's better to have clean commits and 98% bisectability rather than unwieldy commits and 100% bisectability). But you can see if there will be other review comments before it's merged.

Thanks for all the updates. As it stands, this PR has substantial changes and apparently doesn't fit with codefreeze of the 2.6 release in progress, so it'll be merged when the merge window opens.

Then, we'll need to remind mergers that build error of

```
Error: lib/libc/minimal/include/stdlib.h:45:WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - srand
```

is false positive here - this patch **does** implement that function, as part of libc.

yashi commented:
@pfalcon, OK, let's see what the others have to say.  Meanwhile, I've updated a commit log by adding the false positive.

> this PR has substantial changes and apparently doesn't fit with codefreeze of the 2.6 release in progress, so it'll be merged when the merge window opens.

Make sense.  :+1:

Thank you, @pfalcon, for all your help and advice.  I really appreciate it. :pray:

yashi commented:
can we merge this?

yashi commented:
@stephanosio, I saw the MAINTAINERS.yml change and you are now the maintainer of libcs.  Would you mind reviewing this PR? I just rebased to the main. Thanks.

stephanosio commented on lib/libc/Kconfig:
```diff
@@ -138,6 +138,21 @@ config MINIMAL_LIBC_LL_PRINTF
 	  Build with long long printf enabled. This will increase the size of
 	  the image.
 
+config MINIMAL_LIBC_GLOBALS
+	bool "Enables functions require global variables"
+	default y
+	help
+	  Enable the minimal libc's implementation of functions which
+	  implicitly access global/static data, e.g. rand(). Such data
+	  must be put into a memory partition if CONFIG_USERSPACE=y,
+	  and disabling this option may save an entry for application
+	  defining many custom partitions.
+
+	  Say 'N' here if you want to save a memory partition by
+	  disabling the following functions:
+	    - rand
+	    - srand
+
```
Noting that we already have the configs to selectively enable/disable different features of the minimal libc, I think it would be better to add a config to enable the "rand feature" and make that select `NEED_LIBC_MEM_PARTITION` so that the globals can be used for userspace.

For example:
```
config MINIMAL_LIBC_RAND
	bool "Enable minimal libc rand implementation"
	select NEED_LIBC_MEM_PARTITION
...
```

stephanosio replied:
p.s. I see that the PR originally added `MINIMAL_LIBC_RAND`, but later changed to `MINIMAL_LIBC_GLOBALS` because memory protection tests were failing due to the libc memory partition being added and there being insufficient number of MPU regions being available for low end Cortex-M cores.

FWIW, I do not think it warrants reworking the config scheme just to get around that.

The current minimal libc implementation explicitly does not make use of the globals due to the high cost associated with it (i.e. additional memory partition for libc). Noting that, I think it is better to just make any minimal libc features that `select NEED_LIBC_MEM_PARTITION` not enabled by default, so users can enable those features if and only if they need them.

yashi replied:
Reverted back to `CONFIG_MINIMAL_LIBC_RAND` instead of `CONFIG_MINIMAL_LIBC_GLOBALS`.

yashi replied:
And also `CONFIG_MINIMAL_LIBC_RAND` now selects `CONFIG_NEED_LIBC_MEM_PARTITION`.

stephanosio commented on include/sys/libc-hooks.h:
```diff
@@ -67,10 +67,15 @@ extern struct k_mem_partition z_malloc_partition;
-    defined(CONFIG_NEED_LIBC_MEM_PARTITION)
-/* Minimal libc has no globals. We do put the stack canary global in the
- * libc partition since it is not worth placing in a partition of its own.
+#if defined(CONFIG_MINIMAL_LIBC_GLOBALS) || defined(CONFIG_NEWLIB_LIBC) || \
+	defined(CONFIG_STACK_CANARIES) || defined(CONFIG_NEED_LIBC_MEM_PARTITION)
```
`CONFIG_MINIMAL_LIBC_GLOBALS` should not be necessary if we use `CONFIG_NEED_LIBC_MEM_PARTITION` as recommended above.

yashi replied:
done

stephanosio commented on lib/libc/minimal/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ zephyr_library_sources(
   source/stdlib/malloc.c
   source/stdlib/bsearch.c
   source/stdlib/exit.c
+  source/stdlib/rand.c
```
Need to remove this as per the `zephyr_library_sources_ifdef` below.

yashi replied:
done

cfriedt commented:
@yashi - can you please address any remaining comments?

yashi commented:
Yes. Will do it in this weekend.

yashi commented on CMakeLists.txt:
```diff
@@ -989,11 +989,11 @@ if(CONFIG_USERSPACE)
 
-  if(CONFIG_NEWLIB_LIBC)
-    set(NEWLIB_PART -l libc.a z_libc_partition)
+  if(CONFIG_MINIMAL_LIBC OR CONFIG_NEWLIB_LIBC)
+    set(LIBC_PART -l libc.a z_libc_partition)
```
I don't know what I was doing.  But indeed it works without it.  Reverted.

yashi commented on lib/libc/minimal/source/stdlib/rand.c:
```diff
@@ -0,0 +1,29 @@
+#include <sys/libc-hooks.h>
+#include <app_memory/app_memdomain.h>
+
+#define LIBC_DATA K_APP_DMEM(z_libc_partition)
```
@stephanosio, is this the right way to specify the global variable in minimal libc?

stephanosio replied:
Yes. `z_libc_partition` is valid as long as `CONFIG_NEED_LIBC_MEM_PARTITION=y`.

yashi replied:
Thanks.

gmarull commented on doc/guides/porting/arch.rst:
```diff
@@ -795,7 +795,7 @@ details:
 Some architectures may need to update software memory management structures
 or modify hardware registers on another CPU when memory domain APIs are invoked.
-If so, CONFIG_ARCH_MEM_DOMAIN_SYNCHRONOUS_API must be selected by the
+If so, :option:`CONFIG_ARCH_MEM_DOMAIN_SYNCHRONOUS_API` must be selected by the
```
nit: use `:kconfig:` to reference Kconfig options

yashi replied:
done.  Thanks.

stephanosio commented on include/sys/libc-hooks.h:
```diff
@@ -69,8 +69,13 @@ extern struct k_mem_partition z_malloc_partition;
     defined(CONFIG_NEED_LIBC_MEM_PARTITION)
-/* Minimal libc has no globals. We do put the stack canary global in the
- * libc partition since it is not worth placing in a partition of its own.
+/* Some architectures might want to save a memory partition used by
+ * minimal libc.  In fact, many functions in minimal libs doesn't
+ * require any global variables/states. Thus we only use the partition
+ * when functions using the globals are enabled.
+ *
+ * We do put the stack canary global in the libc partition since it is
+ * not worth placing in a partition of its own.
```
I am not sure if I understand this comment.

The scheme post-this PR is as simple as:
1. All newlib globals will be placed into `z_libc_partition`.
2. Minimal C library globals, if any, will be placed into `z_libc_partition`.
2. Stack canary globals will be placed into `z_libc_partition` since it is not worth placing in its own partition.
3. Some architectures may place the global pointer to the thread local storage in `z_libc_partition` since it is not worth placing in its own partition.

yashi replied:
I just copy & pasted your comment. Hope you don't mind.

stephanosio commented on lib/libc/Kconfig:
```diff
@@ -145,6 +145,20 @@ config MINIMAL_LIBC_OPTIMIZE_STRING_FOR_SIZE
 
+config MINIMAL_LIBC_RAND
+	bool "Enables rand and srand functions"
+	default y
```
Should not be `default y` as per https://github.com/zephyrproject-rtos/zephyr/pull/32667#discussion_r656414206.

Any **minimal** C library features that predicate on such a heavy requirement as requiring an additional memory partition should not be enabled by default, and should only be enabled on demand.

yashi replied:
done.  Thanks.

stephanosio reviewed (changes requested):
Some additional comments

yashi commented on doc/guides/porting/arch.rst:
```diff
@@ -795,7 +795,7 @@ details:
 Some architectures may need to update software memory management structures
 or modify hardware registers on another CPU when memory domain APIs are invoked.
-If so, CONFIG_ARCH_MEM_DOMAIN_SYNCHRONOUS_API must be selected by the
+If so, :option:`CONFIG_ARCH_MEM_DOMAIN_SYNCHRONOUS_API` must be selected by the
```
done.  Thanks.

yashi commented on doc/reference/usermode/memory_domain.rst:
```diff
@@ -273,9 +274,9 @@ The build-time logic for setting up automatic memory partitions is in
-For example, if the Newlib C library is enabled, the Newlib globals all need
-to be placed in ``z_libc_partition``. The invocation of the script in the
-top-level ``CMakeLists.txt`` adds the following:
+For example, if the Minimal C library or the Newlib C library is enabled, the
+all libc globals need to be placed in ``z_libc_partition``. The invocation of
+the script in the top-level ``CMakeLists.txt`` adds the following:
```
done.  Thanks.

cfriedt commented:
fixes #6493

stephanosio reviewed (approved):
LGTM
